### PR TITLE
Some icu_calendar cleanups

### PR DESCRIPTION
**Pull request by Manishearth**
From review in https://github.com/unicode-org/icu4x/issues/5632

### CODE REVIEW

File: components/calendar/benches/date.rs [hunk 1]
```diff
@@ -258,7 +258,7 @@ fn date_benches(c: &mut Criterion) {
         &fxs,
         icu::calendar::islamic::IslamicUmmAlQura::new_always_calculating(),
         |y, m, d| {
-            Date::try_new_ummalqura_date(
+            Date::try_new_ummalqura_date_with_calendar(
```
Comment (robertbastian) [inline components/calendar/benches/date.rs:261]: shouldn't it be `try_new_ummalqura_with_calendar`? We don't say `DecimalFormatter::try_new_decimal_formatter_with_any_provider`

Comment (Manishearth) [reply components/calendar/benches/date.rs:261]: good catch! can't believe I didn't notice this

File: README.md [hunk 2]
```diff
@@ -46,7 +46,7 @@ let dtf = NeoFormatter::try_new(
 )
 .expect("locale should be present in compiled data");
 
-let date = DateTime::try_new_iso_datetime(2020, 9, 12, 12, 35, 0).expect("datetime should be valid");
+let date = DateTime::try_new_iso(2020, 9, 12, 12, 35, 0).expect("datetime should be valid");
```
Comment (robertbastian) [inline README.md:49]: Out of scope for this PR, but can we make the calendar types be date factories instead? 

```rust
ISO.try_new_date(2020, 9, 12)
```

This way we'll only have a single `try_new_date` method on the calendar trait, not one name per calendar.

Comment (sffc) [reply README.md:49]: It's nice to not have to import the calendar name. You can just import `Date` and call the fn.

Comment (Manishearth) [reply README.md:49]: yeah, that's the goal of this.

Comment (robertbastian) [reply README.md:49]: This definitely doesn't work with all the `_with_calendar` methods you're adding.

> You can just import Date and call the fn.

You *just* have to parse this list to figure out which fn you need. And of course this list doesn't use the same identifiers as calendar names
image

Comment (Manishearth) [reply README.md:49]: > This definitely doesn't work with all the `_with_calendar` methods you're adding.
> 
> >

We can also add compiled data variants of those functions without `_with_calendar`, that's why they're disambiguated in the first place. I could add that now.

Comment (robertbastian) [reply README.md:49]: That's blowing up that list even more, no?

Comment (Manishearth) [reply README.md:49]: Yes.

Comment (sffc) [reply README.md:49]: I see these as being convenience constructors. They shouldn't, and I think they don't, export functionality that you can't get by using `try_new_from_codes` which is the generic constructor. Right?

I guess I'm not opposed to `Date::<Gregorian>::try_new(...)` where you implement a `try_new` fn for all the calendars, all with the same name and disambiguated by the type param. We can leave `try_new_iso` as a special case.

**Comment by Manishearth** [comment]
Can we land this? The opened issue was marked as out of scope and I have more refactorings to do to this crate.

**Review by robertbastian — DISMISSED** [review]
This wasn't in my review queue.

**Review by sffc — DISMISSED** [review]

**Comment by Manishearth** [comment]
@robertbastian yeah it was more directed to Shane to clarify the issue was a non issue.

Will merge once I fix the changelog issue

**Review by sffc — APPROVED** [review]

**Pull request merged**